User input receives feedback, can quit or restart game

liuwinshen opened the pull request:
This PR is for the following Trello cards:
https://trello.com/c/TaI361iF/5-compare-guess-with-code-25
https://trello.com/c/5m3TaEai/6-give-feedback-to-user-wrong-guess-1
https://trello.com/c/unzXfHdh/7-give-feedback-to-user-win-loss-15
https://trello.com/c/pWel60kE/9-exit-game-options-1

Main changes:
- simplified InputGuess class methods
- moved methods that check guesses (against the code and return red/white pins) from the CodeGenerator class to Mastermind class
- added printouts and replay options for win/loss
- added game commands (quit, restart) to Mastermind class

I would especially love feedback on my Mastermind class because it seems to do too much but I wanted to minimize dependencies and it seemed cleaner to keep these methods all together.

Commit:
Updated Readme

Commit:
Updated Readme

Commit:
refactored tests using  and clearer descriptions

Commit:
refactored tests using  and clearer descriptions

Commit:
successfully compares guess to code and returns red/white pin count

Commit:
successfully compares guess to code and returns red/white pin count

Commit:
simplified input guess validation methods

Commit:
simplified input guess validation methods

Commit:
simplified method to get red and white pins, updated tests

Commit:
simplified method to get red and white pins, updated tests

Commit:
added Play method to loop through prompting user guesses until none remaining

Commit:
added Play method to loop through prompting user guesses until none remaining

Commit:
Play method gives feedback and prints message for win/loss

Commit:
Play method gives feedback and prints message for win/loss

Commit:
fixed indexing bug in #red_and_white and got passing tests

Commit:
fixed indexing bug in #red_and_white and got passing tests

Commit:
renamed InputGuess #input method to #guess for clarity

Commit:
renamed #input method to #guess in InputGuess

Commit:
added quit command, simplified Code::red_and_white

Commit:
moved Code methods to Mastermind, split #play method and renamed to #loop_guesses

Commit:
restart game works but prints out feedback upon restarting after winning

Commit:
fixed post-game feedback bug and all tests pass. game working

pwentz commented on lib/code.rb:
```diff
@@ -1,6 +1,13 @@
+  end
+
+  def secret_code
+    Array.new(4).map { |_| @short_colors.sample }.join
```
😎

pwentz commented on lib/inputguess.rb:
```diff
@@ -1,44 +1,28 @@
+    until valid_guess
+      print "Make a guess using available colors #{colors}: "
+      user_input = clean_guess(gets.chomp)
+      return user_input if user_input == "QUIT" || user_input == "RESTART"
```
Looks like we have some stuff in this `guess` function that we might want to test. Perhaps we can pull this out into a testable function?

pwentz commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,122 @@
+      puts "You have #{@remaining_guesses} guesses remaining."
+      guess = @game_input.guess(short_colors)
+      quit if guess == "QUIT"
+      if guess == "RESTART"
```
Looks like we're duplicating some of the logic inside `InputGuess` class. Maybe we can create a `Guess` class that can take the user input, validate it, and respond to `restart?`, `quit?`, and any other stuff you'd need to find out about the user's guess

liuwinshen replied:
Agreed. Would it work to move quit and restart functionality to `InputGuess` rather than create a new class? It's technically not a guess being input but I could generalize the `InputGuess` class to be `UserInput`.

pwentz replied:
You're definitely on the right track here, but I'd like to see it in a simple class on it's own. In object-oriented programming, we create objects with either data or behavior and we pass these objects around to other objects. Since each object has it's own set of behavior, we can encapsulate logic that pertains to a certain domain/format inside a method on that object.

So a class like `InputGuess` can be a little misleading because it seems to represent a verb, not a noun. A more object oriented approach might be something like the following:
```rb
receiver = InputReceiver.new
text = receiver.get_input # .chomp and other immediate formatting can be encapsulated in this guy
guess = Guess.new(text) # now we can ask the guess whatever we want to know about it
guess.quit?
guess.restart?
guess.to_s
guess.to_a
result = GuessResult.new(guess, some_secret_code)
result.correct?
result.incorrect?
```
An ideal OO application would have all of the business/game logic encapsulated inside different, but focused (as in, they do one thing and do it well) objects which get passed around to other objects. Notice in the example how `GuessResult` takes a `Guess` object. The `GuessResult` can use the convenience methods on the `Guess` object to determine whether it is correct or not! 


We can work on this more tomorrow if this doesn't make sense!

pwentz commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,122 @@
+    guess_color_count = color_count(guess)
+
+    white_count = 0
+    guess_color_count.each do |k, v|
```
[here's a standard library method](https://ruby-doc.org/core-2.4.0/Enumerable.html#method-i-reduce) that can help you clean up logic like this

pwentz commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,122 @@
+        end
+      end
+    end
+    puts "Red pins: #{red_count} \nWhite pins: #{white_count}"
```
I noticed that this gets logged when the tests get run and it makes the output tricky to read. Also, printing to the terminal is something that we don't really need to test (because we know `puts` works). Let's try and separate this printing logic from the logic you want to test.

pwentz commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,122 @@
+    {:red => red_count, :white => white_count}
+  end
+
+  def compare(guess)
```
I think this name here is a bit ambiguous. Think about how someone using your `Mastermind` class would call this:
```rb
mastermind = Mastermind.new(..)
mastermind.compare(...)
```
We're not really sure what we're comparing because there is not much context. If you had this method in a tiny, but focused class called `GuessChecker` or `ScoreKeeper`, then it would make a bit more sense with the extra context. Not a big deal, but something to think about when naming stuff in the future.

pwentz reviewed (commented):
Cursory review looks good. Great improvement overall from the first iteration. I'll can leave more comments later.

Commit:
separated guess check methods to new GuessChecker class with specs

Commit:
fixed failing tests and game restart functionality

liuwinshen commented:
@pwentz I made changes based on your feedback. Main changes:

- separated guess checking functionality to its own class (added tests for it, may remove some redundant ones)
- refactored guess checking to calculate red pins then white (so no more crazy hash logic)
- removed duplicated logic/code from quit and restart commands

Feel free to take a look this week. Thank you!

pwentz commented:
Nice job with these changes. I'm definitely noticing some steady improvement. I made one more code comment that is pretty high-level. Perhaps we could work on some of that tomorrow.

Commit:
separated InputGuess class into Guess and GuessValidator classes

Commit:
refactored GuessValidator to not need several instances per use and fixed tests, InputReceiver and Mastermind still broken

soft-circles commented on lib/game_start.rb:
```diff
@@ -1,15 +1,18 @@
 
-      The number of times each color appears in the code is #{game.color_count(secret_code)}.
-      The number of times each color appears in your guess is #{game.color_count(user_guess.chars)}."
+play
```
It might make sense to wrap this in a `Game` class and call `Game.play`.

soft-circles commented on spec/guess_checker_spec.rb:
```diff
@@ -0,0 +1,110 @@
+        checker.check_red("ROYG", "RGYO")
+        expect(checker.check_code).to eq("-G-O")
+      end
+
```
Try not getting into the habit of pushing up pending tests (they tend to be forgotten ;) )

soft-circles commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,90 @@
       end
-    color_count
+    end
+
```
This is a pretty long method. Can we trim this down a bit?

soft-circles commented on lib/guess_validator.rb:
```diff
@@ -0,0 +1,62 @@
+
+    errors
+  end
+
```
Great work here. We can further clean up this method by encapsulating our `if` statements. You can create a method like `run_validations` and run all your `if` statements in that while just making one call to `run_validations` in your `validate` method. It'll make your method a lot easier to read and understand I think.

Commit:
renamed Guess to UserInput and fixed tests

Commit:
fixed quit, restart, loop until guesses are valid

soft-circles commented on lib/guess_checker.rb:
```diff
@@ -0,0 +1,48 @@
+
+  def check_red(guess, code)
+    red_count = 0
+    guess.length.times do |i|
```
I'd suggest using an Enumerable here, check out `.each_with_index` [here](https://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-each_with_index)

soft-circles commented on lib/guess_checker.rb:
```diff
@@ -0,0 +1,48 @@
+
+  def check_white(guess, code)
+    white_count = 0
+    guess.length.times do |i|
```
I'd suggest using an Enumerable here, check out `.each_with_index` [here](https://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-each_with_index)

soft-circles commented on lib/guess_checker.rb:
```diff
@@ -0,0 +1,48 @@
+  def check_white(guess, code)
+    white_count = 0
+    guess.length.times do |i|
+      if guess[i].match(/[A-Z]/)
```
Why are we doing a regex match here? `check_white` seems to be doing a little too much maybe?

soft-circles commented on lib/guess_checker.rb:
```diff
@@ -0,0 +1,48 @@
+    @check_code = original_code.clone
+    red_count = check_red(@current_guess, @check_code)
+    white_count = check_white(@current_guess, @check_code)
+    @pins = {:red => red_count, :white => white_count}
```
In ruby a shortcut for keys as symbols can be written like so: `{red: red_count, white: white_count}` this is the exact same thing that you've got a written but a little cleaner

soft-circles commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,112 @@
+  end
+
+  def restart
+    puts "New game coming right up."
```
Let's abstract away our messages to another class or module.

soft-circles commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,112 @@
+  end
+
+  def quit
+    puts "So long for now!"
```
Let's abstract away our messages to another class or module.

soft-circles commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,112 @@
+
+  def log_guess(guess)
+    @past_guesses << guess
+    puts "Your previous guesses were #{@past_guesses}."
```
Let's abstract away our messages to another class or module.

soft-circles commented on lib/mastermind.rb:
```diff
@@ -1,40 +1,112 @@
+  end
+
+  def replay
+    puts "Would you like to play again? (y/n) "
```
Let's abstract away our messages to another class or module.

Commit:
split out Mastermind class methods, added GameCommands, Messages, and UserInput modules

Commit:
added feedback and win loss checks, restart does not work yet

Commit:
cleaned up UserInput, still need restart to start new game

Commit:
added #win? tests to Mastermind

Commit:
moved cloning down a level closer to where guess and code strings are modified in GuessChecker

liuwinshen commented:
@pwentz Thank you and @akiraandy for your feedback! 

I simplified methods in my Mastermind class, split out input-related functions into a UserInput module and GuessValidator class, and created modules for GameCommands and Messages. 

I'm stuck trying to get restart game functionality to work again, but otherwise the game is playable and tests are passing. I need more tests for my Mastermind class and would love advice on testing UserInput::get_valid_input.

May you please take a look and let me know what you think? From there, I'd like to merge these changes (perhaps without restart functionality) and continue working on a new branch. Thank you!

Commit:
built-in #load is not working, switched order of system clear and restart_message
